pkg/natdiscovery/request_send.go

```diff
@@ -0,0 +1,152 @@
+/*
+© 2021 Red Hat, Inc. and others
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package natdiscovery
+
+import (
+	"net"
+
+	"github.com/google/gopacket/routing"
+	"github.com/pkg/errors"
+	"google.golang.org/protobuf/proto"
+	"k8s.io/klog"
+
+	natproto "github.com/submariner-io/submariner/pkg/natdiscovery/proto"
+)
+
+func (nd *natDiscovery) sendCheckRequestByRemoteID(id string) error {
+	remoteNAT := nd.remoteEndpoints[id]
+	return nd.sendCheckRequest(remoteNAT)
+}
+
+func (nd *natDiscovery) sendCheckRequest(remoteNAT *RemoteEndpointNAT) error {
+	var errPrivate, errPublic error
+	var reqID uint64
+	if remoteNAT.endpoint.Spec.PrivateIP != "" {
+		reqID, errPrivate = nd.sendCheckRequestToTargetIP(remoteNAT, remoteNAT.endpoint.Spec.PrivateIP)
+		if errPrivate == nil {
+			remoteNAT.lastPrivateIPRequestID = reqID
+		}
+	}
+
+	if remoteNAT.endpoint.Spec.PublicIP != "" {
+		reqID, errPublic = nd.sendCheckRequestToTargetIP(remoteNAT, remoteNAT.endpoint.Spec.PublicIP)
+		if errPublic == nil {
+			remoteNAT.lastPublicIPRequestID = reqID
+		}
+	}
+
+	if errPrivate != nil && errPublic != nil {
+		return errors.Errorf("error happened trying to discover both public & private IPs of endpoint %q, [%s, %s]",
+			remoteNAT.endpoint.Spec.CableName, errPublic, errPrivate)
+	}
+
+	if errPrivate != nil {
+		return errors.Wrapf(errPrivate, "error happened trying to NAT-discover private IP of endpoint %q",
+			remoteNAT.endpoint.Spec.CableName)
+	}
+
+	if errPublic != nil {
+		return errors.Wrapf(errPublic, "error happened trying to NAT-discover public IP of endpoint %q",
+			remoteNAT.endpoint.Spec.CableName)
+	}
+
+	return nil
+}
+
+func (nd *natDiscovery) sendCheckRequestToTargetIP(remoteNAT *RemoteEndpointNAT, targetIP string) (uint64, error) {
+	nd.requestCounter++
+
+	targetPort, err := extractNATDiscoveryPort(remoteNAT.endpoint)
+
+	if err != nil {
+		return 0, err
+	}
+
+	sourceIP, err := nd.findSrcIP(targetIP)
+	if err != nil {
+		klog.Warningf("unable to determine source IP while preparing NAT discovery request to endpoint %q: %s",
```

Comment:
Ideally, this case will not hit since we will have a default route on the host. But in the rare case if we are unable to get sourceIP, should we return here?

Reply:
It happens in my laptop ':D

Apparently because I have the same route & IP on my wifi & ethernet :-)

It's a weird case I guess.

But In that case we still want to send the packet at least and try?

Reply:
When there is an error, will it also return a valid sourceIP? 

If not, we will endup with an empty sourceIP in request.UsingSrc right?

Reply:
Yes that is supported 

> One small clarification, otherwise it looks good.

Yes that's supported (thinking of this case), but we won't be able to detect modified src ip, only modified src port.

https://github.com/submariner-io/submariner/pull/1185/files#diff-2ce27aa4f46793624b7693021b9cdc8ccd90aa88ebf7a2428c4c902e4e844a1fR82